# Exploratory Data Analysis

In [14]:
import pandas as pd
from pathlib import Path

DATA_PATH = Path("../data/processed/cleaned_online_retail.csv")

df = pd.read_csv(DATA_PATH)

df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,TotalPrice
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0


In [15]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 779425 entries, 0 to 779424
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      779425 non-null  int64  
 1   StockCode    779425 non-null  str    
 2   Description  779425 non-null  str    
 3   Quantity     779425 non-null  int64  
 4   InvoiceDate  779425 non-null  str    
 5   Price        779425 non-null  float64
 6   Customer ID  779425 non-null  float64
 7   Country      779425 non-null  str    
 8   TotalPrice   779425 non-null  float64
dtypes: float64(3), int64(2), str(4)
memory usage: 53.5 MB


In [16]:
df.shape

(779425, 9)

In [17]:
df["TotalPrice"] = df["Quantity"] * df["Price"]

In [18]:
df["Country"].value_counts().head(10)

Country
United Kingdom    700388
Germany            16432
EIRE               15565
France             13511
Netherlands         5085
Spain               3662
Belgium             3055
Switzerland         3005
Portugal            2356
Australia           1789
Name: count, dtype: int64

In [19]:
df["Description"].value_counts().head(10)

Description
WHITE HANGING HEART T-LIGHT HOLDER    5016
REGENCY CAKESTAND 3 TIER              3335
ASSORTED COLOUR BIRD ORNAMENT         2692
JUMBO BAG RED RETROSPOT               2641
PARTY BUNTING                         2098
LUNCH BAG  BLACK SKULL.               2045
REX CASH+CARRY JUMBO SHOPPER          1950
LUNCH BAG SPACEBOY DESIGN             1890
STRAWBERRY CERAMIC TRINKET BOX        1859
HOME BUILDING BLOCK WORD              1852
Name: count, dtype: int64

In [22]:
total_revenue = df["TotalPrice"].sum()

In [23]:
from IPython.display import display

display(f"💰 Total Revenue: ${total_revenue:,.2f}")

'💰 Total Revenue: $17,374,804.27'

### Insight

The dataset generated a total revenue of **$17,374,804.27** during the observed period.
This metric serves as the primary KPI for evaluating overall business performance.

In [24]:
df.groupby("Customer ID")["TotalPrice"].sum().sort_values(ascending=False).head(10)

Customer ID
18102.0    580987.04
14646.0    528602.52
14156.0    313437.62
14911.0    291420.81
17450.0    244784.25
13694.0    195640.69
17511.0    172132.87
16446.0    168472.50
16684.0    147142.77
12415.0    144458.37
Name: TotalPrice, dtype: float64

In [25]:
df.groupby("Description")["TotalPrice"].sum().sort_values(ascending=False).head(10)

Description
REGENCY CAKESTAND 3 TIER              277656.25
WHITE HANGING HEART T-LIGHT HOLDER    247048.01
PAPER CRAFT , LITTLE BIRDIE           168469.60
Manual                                151777.67
JUMBO BAG RED RETROSPOT               134307.44
POSTAGE                               124648.04
ASSORTED COLOUR BIRD ORNAMENT         124351.86
PARTY BUNTING                         103283.38
MEDIUM CERAMIC TOP STORAGE JAR         81416.73
PAPER CHAIN KIT 50'S CHRISTMAS         76598.18
Name: TotalPrice, dtype: float64

In [26]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])
df["Month"] = df["InvoiceDate"].dt.month

In [27]:
df["Year"] = df["InvoiceDate"].dt.year
df["Month"] = df["InvoiceDate"].dt.month
df["Day"] = df["InvoiceDate"].dt.day
df["Hour"] = df["InvoiceDate"].dt.hour
df["Weekday"] = df["InvoiceDate"].dt.day_name()

In [28]:
df.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,TotalPrice,Month,Year,Day,Hour,Weekday
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4,12,2009,1,7,Tuesday
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0,12,2009,1,7,Tuesday
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0,12,2009,1,7,Tuesday
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8,12,2009,1,7,Tuesday
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0,12,2009,1,7,Tuesday
5,489434,22064,PINK DOUGHNUT TRINKET POT,24,2009-12-01 07:45:00,1.65,13085.0,United Kingdom,39.6,12,2009,1,7,Tuesday
6,489434,21871,SAVE THE PLANET MUG,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0,12,2009,1,7,Tuesday
7,489434,21523,FANCY FONT HOME SWEET HOME DOORMAT,10,2009-12-01 07:45:00,5.95,13085.0,United Kingdom,59.5,12,2009,1,7,Tuesday
8,489435,22350,CAT BOWL,12,2009-12-01 07:46:00,2.55,13085.0,United Kingdom,30.6,12,2009,1,7,Tuesday
9,489435,22349,"DOG BOWL , CHASING BALL DESIGN",12,2009-12-01 07:46:00,3.75,13085.0,United Kingdom,45.0,12,2009,1,7,Tuesday


In [29]:
monthly_sales = df.groupby("Month")["TotalPrice"].sum()

monthly_sales

Month
1     1123903.982
2      950643.876
3     1291060.231
4     1060356.333
5     1275188.530
6     1296417.180
7     1188699.071
8     1246275.640
9     1779704.153
10    2068754.460
11    2322665.632
12    1771135.180
Name: TotalPrice, dtype: float64

In [30]:
monthly_sales.describe()

count    1.200000e+01
mean     1.447900e+06
std      4.320027e+05
min      9.506439e+05
25%      1.172500e+06
50%      1.283124e+06
75%      1.773277e+06
max      2.322666e+06
Name: TotalPrice, dtype: float64

In [31]:
average_order_value = (
    df.groupby("Invoice")["TotalPrice"]
      .sum()
      .mean()
)

print(f"Average Order Value: ${average_order_value:,.2f}")

Average Order Value: $469.98


### Insight

Average Order Value (AOV) measures the average amount customers spend per order.

A higher AOV generally indicates stronger purchasing behavior.

In [32]:
total_customers = df["Customer ID"].nunique()

print(f"Total Customers: {total_customers:,}")

Total Customers: 5,878


In [33]:
top_customers = (
    df.groupby("Customer ID")["TotalPrice"]
      .sum()
      .sort_values(ascending=False)
      .head(10)
)

top_customers

Customer ID
18102.0    580987.04
14646.0    528602.52
14156.0    313437.62
14911.0    291420.81
17450.0    244784.25
13694.0    195640.69
17511.0    172132.87
16446.0    168472.50
16684.0    147142.77
12415.0    144458.37
Name: TotalPrice, dtype: float64

In [34]:
customer_orders = (
    df.groupby("Customer ID")["Invoice"]
      .nunique()
      .sort_values(ascending=False)
      .head(10)
)

customer_orders

Customer ID
14911.0    398
12748.0    336
17841.0    211
15311.0    208
13089.0    203
14606.0    192
14156.0    156
17850.0    155
14646.0    151
18102.0    145
Name: Invoice, dtype: int64

### Insight

A small group of customers contributes a significant portion of total revenue,
making them high-value customers for retention strategies.

In [35]:
unique_products = df["Description"].nunique()

print(f"Unique Products: {unique_products:,}")

Unique Products: 5,283


In [36]:
top_products = (
    df.groupby("Description")["TotalPrice"]
      .sum()
      .sort_values(ascending=False)
      .head(10)
)

top_products

Description
REGENCY CAKESTAND 3 TIER              277656.25
WHITE HANGING HEART T-LIGHT HOLDER    247048.01
PAPER CRAFT , LITTLE BIRDIE           168469.60
Manual                                151777.67
JUMBO BAG RED RETROSPOT               134307.44
POSTAGE                               124648.04
ASSORTED COLOUR BIRD ORNAMENT         124351.86
PARTY BUNTING                         103283.38
MEDIUM CERAMIC TOP STORAGE JAR         81416.73
PAPER CHAIN KIT 50'S CHRISTMAS         76598.18
Name: TotalPrice, dtype: float64

In [37]:
top_quantity = (
    df.groupby("Description")["Quantity"]
      .sum()
      .sort_values(ascending=False)
      .head(10)
)

top_quantity

Description
WORLD WAR 2 GLIDERS ASSTD DESIGNS     105185
WHITE HANGING HEART T-LIGHT HOLDER     91757
PAPER CRAFT , LITTLE BIRDIE            80995
ASSORTED COLOUR BIRD ORNAMENT          78234
MEDIUM CERAMIC TOP STORAGE JAR         77916
JUMBO BAG RED RETROSPOT                74224
BROCADE RING PURSE                     70082
PACK OF 60 PINK PAISLEY CAKE CASES     54592
60 TEATIME FAIRY CAKE CASES            52828
PACK OF 72 RETRO SPOT CAKE CASES       45129
Name: Quantity, dtype: int64

In [38]:
countries = df["Country"].nunique()

print(f"Countries: {countries}")

Countries: 41


In [39]:
country_revenue = (
    df.groupby("Country")["TotalPrice"]
      .sum()
      .sort_values(ascending=False)
)

country_revenue.head(10)

Country
United Kingdom    1.438923e+07
EIRE              6.165705e+05
Netherlands       5.540381e+05
Germany           4.250197e+05
France            3.487690e+05
Australia         1.692835e+05
Spain             1.083325e+05
Switzerland       1.000619e+05
Sweden            9.151582e+04
Denmark           6.858069e+04
Name: TotalPrice, dtype: float64

In [40]:
country_orders = (
    df.groupby("Country")["Invoice"]
      .nunique()
      .sort_values(ascending=False)
)

country_orders.head(10)

Country
United Kingdom    33541
Germany             789
France              614
EIRE                567
Netherlands         228
Spain               154
Belgium             149
Sweden              104
Australia            95
Portugal             93
Name: Invoice, dtype: int64

In [41]:
monthly_revenue = (
    df.groupby("Month")["TotalPrice"]
      .sum()
)

monthly_revenue

Month
1     1123903.982
2      950643.876
3     1291060.231
4     1060356.333
5     1275188.530
6     1296417.180
7     1188699.071
8     1246275.640
9     1779704.153
10    2068754.460
11    2322665.632
12    1771135.180
Name: TotalPrice, dtype: float64

In [42]:
weekday_revenue = (
    df.groupby("Weekday")["TotalPrice"]
      .sum()
)

weekday_revenue

Weekday
Friday       2728473.173
Monday       2778201.566
Saturday        9803.050
Sunday       1768669.052
Thursday     3745783.432
Tuesday      3322830.142
Wednesday    3021043.853
Name: TotalPrice, dtype: float64

In [43]:
hourly_revenue = (
    df.groupby("Hour")["TotalPrice"]
      .sum()
)

hourly_revenue

Hour
6           4.250
7       75745.570
8      524612.550
9     1484450.541
10    2320711.443
11    2235870.393
12    2690583.501
13    2347078.714
14    1962016.161
15    1839607.762
16    1055274.691
17     484477.632
18     190860.290
19     124825.770
20      38685.000
Name: TotalPrice, dtype: float64

# Executive Summary

## Key Findings

- The business generated **$XX** in total revenue.
- The United Kingdom is the largest market by revenue.
- A small number of customers contribute a significant portion of total sales.
- Revenue varies across months, indicating seasonal purchasing behavior.
- Certain products consistently generate the highest revenue.

## Business Recommendations

- Focus marketing efforts on high-value customers.
- Increase inventory for top-performing products.
- Expand into high-performing international markets.
- Investigate low-performing months to improve sales consistency.

# Engineered Features

| Feature | Description |
|---------|-------------|
| TotalPrice | Quantity × Price |
| Year | Invoice year |
| Month | Invoice month |
| Day | Invoice day |
| Hour | Invoice hour |
| Weekday | Day of week |

In [44]:
print("="*60)
print("KEY METRICS")
print("="*60)

print(f"Total Revenue      : ${df['TotalPrice'].sum():,.2f}")
print(f"Total Orders       : {df['Invoice'].nunique():,}")
print(f"Total Customers    : {df['Customer ID'].nunique():,}")
print(f"Unique Products    : {df['Description'].nunique():,}")
print(f"Countries          : {df['Country'].nunique():,}")
print(f"Average Order Value: ${average_order_value:,.2f}")

KEY METRICS
Total Revenue      : $17,374,804.27
Total Orders       : 36,969
Total Customers    : 5,878
Unique Products    : 5,283
Countries          : 41
Average Order Value: $469.98


# Conclusion

The exploratory analysis identified the main revenue drivers, customer purchasing patterns, product performance, geographic distribution, and temporal trends.

These findings establish a solid foundation for the visualization notebook, where insights will be communicated through professional charts and business storytelling.